File: references/java-sdk/src/main/java/blue/language/sdk/LinkedAccessBuilder.java
package blue.language.sdk;

import blue.language.sdk.internal.StepsBuilder;
import blue.language.types.myos.LinkedDocumentsPermissionGranted;
import blue.language.types.myos.LinkedDocumentsPermissionRejected;
import blue.language.types.myos.LinkedDocumentsPermissionRevoked;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public final class LinkedAccessBuilder<P extends DocBuilder<P>> {

    private final P parent;
    private final String name;
    private Object targetSessionId;
    private String onBehalfOfChannel;
    private String statusPath;
    private final Map<String, LinkBuilder<P>> links = new LinkedHashMap<String, LinkBuilder<P>>();

    private AccessConfig.PermissionTiming permissionTiming = AccessConfig.PermissionTiming.ON_INIT;
    private Class<?> permissionTriggerEventClass;
    private String permissionTriggerDocPath;

    LinkedAccessBuilder(P parent, String name) {
        if (parent == null) {
            throw new IllegalArgumentException("parent is required");
        }
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("linked access name is required");
        }
        this.parent = parent;
        this.name = name.trim();
    }

    public LinkedAccessBuilder<P> targetSessionId(Object targetSessionId) {
        this.targetSessionId = targetSessionId;
        return this;
    }

    public LinkedAccessBuilder<P> onBehalfOf(String channelKey) {
        this.onBehalfOfChannel = channelKey;
        return this;
    }

    public LinkedAccessBuilder<P> statusPath(String statusPath) {
        this.statusPath = statusPath;
        return this;
    }

    public LinkBuilder<P> link(String linkName) {
        if (linkName == null || linkName.trim().isEmpty()) {
            throw new IllegalArgumentException("link name is required");
        }
        String normalized = linkName.trim();
        LinkBuilder<P> builder = links.get(normalized);
        if (builder == null) {
            builder = new LinkBuilder<P>(this, normalized);
            links.put(normalized, builder);
        }
        return builder;
    }

    public LinkedAccessBuilder<P> requestPermissionOnInit() {
        this.permissionTiming = AccessConfig.PermissionTiming.ON_INIT;
        this.permissionTriggerEventClass = null;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public LinkedAccessBuilder<P> requestPermissionOnEvent(Class<?> eventTypeClass) {
        if (eventTypeClass == null) {
            throw new IllegalArgumentException("eventTypeClass is required");
        }
        this.permissionTiming = AccessConfig.PermissionTiming.ON_EVENT;
        this.permissionTriggerEventClass = eventTypeClass;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public LinkedAccessBuilder<P> requestPermissionOnDocChange(String path) {
        if (path == null || path.trim().isEmpty()) {
            throw new IllegalArgumentException("path is required");
        }
        this.permissionTiming = AccessConfig.PermissionTiming.ON_DOC_CHANGE;
        this.permissionTriggerDocPath = path.trim();
        this.permissionTriggerEventClass = null;
        return this;
    }

    public LinkedAccessBuilder<P> requestPermissionManually() {
        this.permissionTiming = AccessConfig.PermissionTiming.MANUAL;
        this.permissionTriggerEventClass = null;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public P done() {
        validate();

        String token = DocBuilder.tokenOf(name, "LINKEDACCESS");
        String requestId = "REQ_LINKED_ACCESS_" + token;

        Map<String, LinkedAccessConfig.LinkConfig> linkConfigs = new LinkedHashMap<String, LinkedAccessConfig.LinkConfig>();
        for (Map.Entry<String, LinkBuilder<P>> entry : links.entrySet()) {
            LinkBuilder<P> builder = entry.getValue();
            linkConfigs.put(entry.getKey(), new LinkedAccessConfig.LinkConfig(builder.read, builder.operations));
        }

        LinkedAccessConfig config = new LinkedAccessConfig(
                name,
                token,
                targetSessionId,
                onBehalfOfChannel,
                requestId,
                statusPath,
                linkConfigs,
                permissionTiming,
                permissionTriggerEventClass,
                permissionTriggerDocPath);
        parent.registerLinkedAccessConfig(config);

        if (config.statusPath() != null) {
            parent.field(config.statusPath(), "pending");
        }

        if (config.permissionTiming() != AccessConfig.PermissionTiming.MANUAL) {
            String workflowKey = "linkedAccess" + token + "RequestPermission";
            Consumer<StepsBuilder> permissionStep = steps -> steps.myOs().requestLinkedDocsPermission(
                    config.onBehalfOfChannel(),
                    config.requestId(),
                    config.targetSessionId(),
                    config.linksAsNodeMap());
            switch (config.permissionTiming()) {
                case ON_INIT:
                    parent.onInit(workflowKey, permissionStep);
                    break;
                case ON_EVENT:
                    parent.onEvent(workflowKey, config.permissionTriggerEventClass(), permissionStep);
                    break;
                case ON_DOC_CHANGE:
                    parent.onDocChange(workflowKey, config.permissionTriggerDocPath(), permissionStep);
                    break;
                default:
                    break;
            }
        }

        parent.onMyOsResponse("linkedAccess" + token + "Granted",
                LinkedDocumentsPermissionGranted.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkLinkedAccessGranted", config.statusPath(), "granted");
                    }
                });

        parent.onMyOsResponse("linkedAccess" + token + "Rejected",
                LinkedDocumentsPermissionRejected.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkLinkedAccessRejected", config.statusPath(), "rejected");
                    }
                });

        parent.onMyOsResponse("linkedAccess" + token + "Revoked",
                LinkedDocumentsPermissionRevoked.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkLinkedAccessRevoked", config.statusPath(), "revoked");
                    }
                });

        return parent;
    }

    private void validate() {
        if (targetSessionId == null) {
            throw new IllegalArgumentException("accessLinked('" + name + "'): targetSessionId is required");
        }
        if (onBehalfOfChannel == null || onBehalfOfChannel.trim().isEmpty()) {
            throw new IllegalArgumentException("accessLinked('" + name + "'): onBehalfOf is required");
        }
        if (links.isEmpty()) {
            throw new IllegalArgumentException("accessLinked('" + name + "'): at least one link(...) is required");
        }
    }

    public static final class LinkBuilder<P extends DocBuilder<P>> {
        private final LinkedAccessBuilder<P> parent;
        private final String linkName;
        private boolean read;
        private final List<String> operations = new ArrayList<String>();

        private LinkBuilder(LinkedAccessBuilder<P> parent, String linkName) {
            this.parent = parent;
            this.linkName = linkName;
        }

        public LinkBuilder<P> read(boolean read) {
            this.read = read;
            return this;
        }

        public LinkBuilder<P> operations(String... operations) {
            if (operations != null) {
                this.operations.addAll(Arrays.asList(operations));
            }
            return this;
        }

        public LinkedAccessBuilder<P> done() {
            parent.links.put(linkName, this);
            return parent;
        }
    }
}

File: libs/dsl-sdk/src/lib/interactions/linked-access-builder.ts
import { toTypeAlias, type TypeLike } from '../core/type-alias.js';
import type {
  LinkedAccessConfig,
  LinkedAccessConfigRegistrationHost,
  LinkedAccessLinkConfig,
  InteractionPermissionTiming,
} from './types.js';

function requireText(value: string | undefined, field: string): string {
  const normalized = value?.trim();
  if (!normalized) {
    throw new Error(`${field} is required`);
  }
  return normalized;
}

function token(name: string): string {
  const normalized = name
    .replace(/[^A-Za-z0-9]+/gu, '_')
    .replace(/^_+|_+$/gu, '')
    .toUpperCase();
  return normalized.length > 0 ? normalized : 'LINKED_ACCESS';
}

type StepsCustomizer = (steps: {
  myOs: (adminChannelKey?: string) => {
    requestLinkedDocsPermission: (
      onBehalfOf: string,
      requestId: string,
      targetSessionId: string,
      linkedPermissions: Record<string, unknown>,
    ) => unknown;
  };
  replaceValue: (name: string, path: string, value: unknown) => unknown;
}) => void;

type DocBuilderLike = {
  field(path: string, value: unknown): unknown;
  onInit(workflowKey: string, customizer: StepsCustomizer): unknown;
  onEvent(
    workflowKey: string,
    eventType: string,
    customizer: StepsCustomizer,
  ): unknown;
  onDocChange(
    workflowKey: string,
    path: string,
    customizer: StepsCustomizer,
  ): unknown;
  onMyOsResponse(
    workflowKey: string,
    responseType: string,
    requestId: string,
    customizer: StepsCustomizer,
  ): unknown;
};

interface LinkedAccessLinkDraft {
  read: boolean;
  operations: string[];
}

export class LinkedAccessBuilder<P> {
  private onBehalfOfValue: string | undefined;
  private targetSessionIdValue: string | undefined;
  private requestIdValue: string | undefined;
  private subscriptionIdValue: string | undefined;
  private statusPathValue: string | undefined;
  private readonly linksValue = new Map<string, LinkedAccessLinkDraft>();
  private permissionTimingValue: InteractionPermissionTiming = 'onInit';
  private permissionTriggerEventTypeValue: string | undefined;
  private permissionTriggerDocPathValue: string | undefined;
  private readonly subscriptionEventsValue: string[] = [];

  constructor(
    private readonly parent: LinkedAccessConfigRegistrationHost<P>,
    private readonly name: string,
  ) {}

  onBehalfOf(channelKey: string): this {
    this.onBehalfOfValue = requireText(channelKey, 'onBehalfOf');
    return this;
  }

  permissionFrom(channelKey: string): this {
    this.onBehalfOf(channelKey);
    return this;
  }

  targetSessionId(sessionId: string): this {
    this.targetSessionIdValue = requireText(sessionId, 'targetSessionId');
    return this;
  }

  requestId(requestId: string): this {
    this.requestIdValue = requireText(requestId, 'requestId');
    return this;
  }

  subscriptionId(subscriptionId: string): this {
    this.subscriptionIdValue = requireText(subscriptionId, 'subscriptionId');
    return this;
  }

  statusPath(statusPath: string): this {
    this.statusPathValue = requireText(statusPath, 'statusPath');
    return this;
  }

  link(name: string): LinkedAccessLinkBuilder<P> {
    const normalizedName = requireText(name, 'link name');
    const draft = this.linksValue.get(normalizedName) ?? {
      read: true,
      operations: [],
    };
    this.linksValue.set(normalizedName, draft);
    return new LinkedAccessLinkBuilder(this, draft);
  }

  requestPermissionOnInit(): this {
    this.permissionTimingValue = 'onInit';
    this.permissionTriggerEventTypeValue = undefined;
    this.permissionTriggerDocPathValue = undefined;
    return this;
  }

  requestPermissionOnEvent(eventType: TypeLike): this {
    this.permissionTimingValue = 'onEvent';
    this.permissionTriggerEventTypeValue = toTypeAlias(eventType);
    this.permissionTriggerDocPathValue = undefined;
    return this;
  }

  requestPermissionOnDocChange(path: string): this {
    this.permissionTimingValue = 'onDocChange';
    this.permissionTriggerDocPathValue = requireText(path, 'permission path');
    this.permissionTriggerEventTypeValue = undefined;
    return this;
  }

  requestPermissionManually(): this {
    this.permissionTimingValue = 'manual';
    this.permissionTriggerDocPathValue = undefined;
    this.permissionTriggerEventTypeValue = undefined;
    return this;
  }

  subscriptionEvents(...eventTypes: TypeLike[]): this {
    this.subscriptionEventsValue.length = 0;
    this.subscriptionEventsValue.push(
      ...eventTypes.map((eventType) => toTypeAlias(eventType)),
    );
    return this;
  }

  done(): P {
    const baseToken = token(this.name);
    if (this.linksValue.size === 0) {
      throw new Error('At least one linked access link must be configured');
    }
    const links: Record<string, LinkedAccessLinkConfig> = {};
    for (const [linkName, linkValue] of this.linksValue.entries()) {
      links[linkName] = {
        read: linkValue.read,
        operations: [...new Set(linkValue.operations)],
      };
    }
    const config: LinkedAccessConfig = {
      name: this.name,
      token: baseToken,
      targetSessionId: requireText(
        this.targetSessionIdValue,
        'targetSessionId',
      ),
      onBehalfOf: requireText(this.onBehalfOfValue, 'onBehalfOf'),
      requestId: this.requestIdValue ?? `REQ_LINKED_${baseToken}`,
      subscriptionId: this.subscriptionIdValue ?? `SUB_LINKED_${baseToken}`,
      ...(this.statusPathValue ? { statusPath: this.statusPathValue } : {}),
      links,
      permissionTiming: this.permissionTimingValue,
      ...(this.permissionTriggerEventTypeValue
        ? { permissionTriggerEventType: this.permissionTriggerEventTypeValue }
        : {}),
      ...(this.permissionTriggerDocPathValue
        ? { permissionTriggerDocPath: this.permissionTriggerDocPathValue }
        : {}),
      permissionFrom: this.onBehalfOfValue,
      subscriptionEvents: [...this.subscriptionEventsValue],
    };
    const registered = this.parent.registerLinkedAccessConfig(config);
    this.applyAutoWiring(config, registered as unknown as DocBuilderLike);
    return registered;
  }

  private applyAutoWiring(
    config: LinkedAccessConfig,
    parent: DocBuilderLike,
  ): void {
    if (config.statusPath) {
      parent.field(config.statusPath, 'pending');
    }

    const workflowPrefix = `linkedAccess${config.token}`;
    const requestPermissionWorkflow: StepsCustomizer = (steps) => {
      const linkedPermissions: Record<string, unknown> = {};
      for (const [linkName, linkConfig] of Object.entries(config.links)) {
        const permission: Record<string, unknown> = {};
        if (linkConfig.read) {
          permission.read = true;
        }
        if (linkConfig.operations.length > 0) {
          permission.singleOps = [...linkConfig.operations];
        }
        linkedPermissions[linkName] = permission;
      }
      steps
        .myOs()
        .requestLinkedDocsPermission(
          config.onBehalfOf,
          config.requestId,
          config.targetSessionId,
          linkedPermissions,
        );
    };

    if (config.permissionTiming === 'onInit') {
      parent.onInit(
        `${workflowPrefix}RequestPermission`,
        requestPermissionWorkflow,
      );
    } else if (config.permissionTiming === 'onEvent') {
      parent.onEvent(
        `${workflowPrefix}RequestPermission`,
        config.permissionTriggerEventType ?? 'Conversation/Event',
        requestPermissionWorkflow,
      );
    } else if (config.permissionTiming === 'onDocChange') {
      parent.onDocChange(
        `${workflowPrefix}RequestPermission`,
        config.permissionTriggerDocPath ?? '/',
        requestPermissionWorkflow,
      );
    }

    parent.onMyOsResponse(
      `${workflowPrefix}Granted`,
      'MyOS/Linked Documents Permission Granted',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkLinkedAccessGranted',
            config.statusPath,
            'granted',
          );
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}Rejected`,
      'MyOS/Linked Documents Permission Rejected',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkLinkedAccessRejected',
            config.statusPath,
            'rejected',
          );
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}Revoked`,
      'MyOS/Linked Documents Permission Revoked',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkLinkedAccessRevoked',
            config.statusPath,
            'revoked',
          );
        }
      },
    );
  }
}

class LinkedAccessLinkBuilder<P> {
  constructor(
    private readonly parent: LinkedAccessBuilder<P>,
    private readonly draft: LinkedAccessLinkDraft,
  ) {}

  read(read: boolean): this {
    if (read === false) {
      throw new Error(
        'accessLinked(...).link(...).read(false) is not supported; MyOS linked-document permissions require read=true',
      );
    }
    this.draft.read = read;
    return this;
  }

  operations(...operations: string[]): this {
    this.draft.operations = operations
      .map((operation) => operation?.trim())
      .filter((operation): operation is string => Boolean(operation));
    return this;
  }

  done(): LinkedAccessBuilder<P> {
    return this.parent;
  }
}
